sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package expansionx contains utilities for starting expansion services for
+// cross-language transforms.
+package expansionx
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+)
+
+// URL is a type used to differentiate between strings and URLs
+type URL string
+
+const (
+	apacheRepository = URL("https://repo.maven.apache.org/maven2")
+	beamGroupID      = "org/apache/beam"
+	jarCache         = "~/.apache_beam/cache/jars"
+)
+
+// GetBeamJar checks a temporary directory for the desired Beam JAR, downloads the
+// appropriate JAR from Maven if not present, then returns the file path to the
+// JAR.
+func GetBeamJar(gradleTarget, version string) (string, error) {
```

Comment:
While it's unlikely to matter too much, for making code like this (which refers to constants and such) more testable, consider making a struct to hold the constants, and the rest to be methods.

This way, one can substitute the constants out in tests (without needing to go for approaches like re-assigning variables, which can be brittle and non-hermetic). Eg with local addresses.

In particular, it would allow for something I suggested to you offline: Being able to fully test this with a fake in process HTTP server. That would be going the extra mile on testing, and I don't expect that in this instance.

The exported function `GetBeamJar` can still be a package level function, but then it's doing something like this.

```
func GetBeamJar(gradleTarget, version string) (string, error) {
  return cacher{
    repo: apacheRepository,
    groupID:   beamGroupID,
    jarCache: jarCache,
  }.getJar(gradleTarget, version)
}
``` 

With all the implementation handled in the getJar method.

Reply:
I like that pattern, I'm going to work on refactoring that and handling tests now (seeing as we don't currently hit code cov on this PR.)

Reply:
Done